### PR TITLE
feat(core): rename async to waitForAsync to avoid confusing

### PR DESCRIPTION
**Pull request by JiaLiPassion**
@angular/core/testing provide `async` test utility, but the name `async` is
confusing with the javascript keyword `async`. And in some test case, if you
want to use both the `async` from `@angular/core/testing` and `async/await`,
you may have to write the code like this.

```typescript
function asyncFunc(): Promise<{value: number}> {
    return new Promise(res => setTimeout(() => { res({value: 1}); }));
  }

  function timeout(obj) {
    setTimeout(() => {
      obj.value += 1;
    });
  }

  it('test async', async () => {
    const r: {value: number} = await asyncFunc();
    timeout(r);
    setTimeout(() => {
      expect(r.value).toBe(2);
    });
   });
```

So in this PR, the `async` is renamed to `asyncTest`.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

### CODE REVIEW

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Reviewed-for: fw-upgrade

Don't forget to update: https://github.com/angular/angular/blob/efdf347f9c04e19c8c4fa8b8c05f4310f42728d5/aio/content/guide/upgrade-setup.md

Otherwise LGTM

And I believe you are also going to rename `fakeAsync()` too.

**Comment by clydin** [comment]
Can you provide some details on when a developer would need to use this:
```
it('test async operations', async(async() => {
  const result = await asyncMethod();
  expect(result).toEqual('expected');
}));
```
instead of this:
```
it('test async operations', async () => {
  const result = await asyncMethod();
  expect(result).toEqual('expected');
});
```

**Comment by jelbourn** [comment]
@JiaLiPassion I'm personally missing some context on this; is this something that's been previously proposed/discussed with the larger team?

**Comment by JiaLiPassion** [comment]
@jelbourn, there is no discussion about this one, I just want to make a proposal by making this PR so it may easier to review and discuss. Thanks.

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: fw-upgrade

**Review by atscott — APPROVED** [review]
reviewed for: fw-router

This would need to land in a major release and should have a migration with it, right?

**Review by atscott — APPROVED** [review]
reviewed-for: fw-router

**Review by pkozlowski-opensource — APPROVED** [review]
The change itself LGTM and I see why we would like to rename those functions. If I understand correctly for now we are just deprecating the `async` / `fakeAsync` functions _and_ removing the usage from the main repo? It would have been slightly more clear with 2 separate commits (one adding a new function + depreciation and another one with usage changes).

Having said all this I guess we will have to plan the actual removal very carefully and be sure to have an automatic migration - otherwise it would be too much of the breaking change (as we can see in the main repo).

Reviewed-for: public-api 
Reviewed-for: fw-core 
Reviewed-for: fw-testing

**Comment by pkozlowski-opensource** [comment]
Also @JiaLiPassion for some reason CI didn't run all the jobs so I'm adding the cleanup label.

**Comment by JiaLiPassion** [comment]
@pkozlowski-opensource , thank you for the review, I have rebased and also split the PR into two commits.

> Having said all this I guess we will have to plan the actual removal very carefully and be sure to have an automatic migration - otherwise it would be too much of the breaking change (as we can see in the main repo).

Yeah, I agree, will try to figure out how to make an automatic migration tool later.

**Comment by clydin** [comment]
I believe this change has broader implications than just the FW changes.  For example, tooling will also be affected (both Angular and third-party).  Instructional material such as courseware, videos, blogs/articles, etc. will need to be updated by their providers as well (this can be a time consuming and expensive endeavor).   As mentioned, a migration would also be of importance as developers would receive potentially hundreds of deprecation warnings after upgrading.

I think it may also be useful to have a larger discussion on whether the `async` helper function should continue to exist moving forward.  And if it should not then a renaming may not be necessary (`fakeAsync` is fairly distinctly named in relation to the `async` keyword).

**Comment by jelbourn** [comment]
@StephenFluin and @mgechev I'd like to get your input on this change and the naming 

TL;DR: having a function named `async` is problematic when there's a native `async` keyword, so this change proposes deprecating the names `async` and `fakeAsync` in favor of `asyncTest` and `fakeAsyncTest`

**Comment by alfaproject** [comment]
I'm with @clydin here. We haven't used `async` ever since we could use native `async/await`. Can't it just be straight up deprecated?

**Comment by shairez** [comment]
I get why `async` needs to change (or to get deprecated).

From my experience - I really didn't see a lot of usage for it as well.

But `fakeAsync` is the more commonly used function and I agree with @clydin that this change will have a huge negative impact in multiple aspects.. while not super necessary since there aren't any collisions with native functions. 

So... I say - do what ever you want with `async`, just [leave `fakeAsync` alone](https://www.youtube.com/watch?v=0wG_h3U568w)  😄

**Comment by yjaaidi** [comment]
+1 to get rid of `async()` somehow :)
The confusing thing with `asyncTest` is that it sounds weird if used with `beforeEach` or `afterEach` (e.g. `beforeEach(asyncTest(...))` ) as we are not talking about a test here.
Maybe `asyncRun` or `asyncWait`? or the opposite `runAsync` or `waitAsync`?

**Comment by yjaaidi** [comment]
By the way, would you agree to replace the component's test schematic:
```typescript
beforeEach(async(() => {
  TestBed.configureTestingModule({})
  .compileComponents();
}));
```
by:
```typescript
beforeEach(async () => {
  await TestBed.configureTestingModule({ })
  .compileComponents();
});
```

this makes tests zoneless friendly by default.


_We might also think about the option below but I am not in favor of it as developers might replace it with a block and forget to add `async / await`._
```typescript
beforeEach(() => TestBed.configureTestingModule({}).compileComponents());
```

cc. @BioPhoton @LayZeeDK

**Comment by petebacondarwin** [comment]
Actually I really like this combo: `runAsync` and `fakeAsync` that would resolve the concerns raised by @clydin , @shairez and @yjaaidi.

**Comment by LayZeeDK** [comment]
I think both `async` and `asyncTest` are equally poor function names for what this one does.

Like many other Zone.js-related things in Angular, this one is pretty difficult to explain and reason about. If I recall correctly, `async` intercepts all monkey-patched APIs through a zone and flushes them after callback execution.

A more descriptive name might be something like `interceptAndFlushAsync` or `waitForAsync`.

@yjaaidi
Regarding the spec generation schematics, I remember reading somewhere that it's not even necessary to run `TestBed#compileComponents` asynchronously if we're using the Angular CLI.

That it's called the same as a keyword has been an issue since Angular version 2.0 since TypeScript has had `async-await` since version 1.7. I'm with @clydin in that this will deprecate a lot of material including Angular.io's testing guide, Angular's codebase and the codebases of a ton of lot of libraries in the ecosystem.

I would love to get a more descriptive name though.

One other aspect that @yjaaidi brings up is related to the flaky future of Zone.js. Angular appications can't target `es2018` as TypeScript output as long as they're using Zone.js unless we force downleveling of `async-await` into the Angular compiler or add an explicit NgZone trigger after each `async-await` statement (also using the compiler). This would need to affect all code imported from `node_modules` as well.

If removing Zone.js from Angular is within scope of +3 major versions (I don't know the roadmap for Angular+Zone.js), would making a big change like this, preceded by a deprecation period, be worth it, seeing that this is only related to NgZone-enabled applications?

**Comment by JiaLiPassion** [comment]
@LayZeeDK, I agree with your points that `async` and `asyncTest` may not be the good names.

> If removing Zone.js from Angular is within scope of +3 major versions (I don't know the roadmap for Angular+Zone.js), would making a big change like this, preceded by a deprecation period, be worth it, seeing that this is only related to NgZone-enabled applications?

`async` and `fakeAsync` are just test helper functions, they have nothing to do with `ngZone`, `async` will call `done` automatically when all `macroTasks and microTasks` finished, and `fakeAsync` just like another version of `jasmine clock and mockDate` to make async test to sync. those functions are using `zone.js` but even Angular doesn't depend on `zone.js` to trigger change detection any longer, the test helper functions still work as expected.

File: packages/core/testing/src/async.ts [hunk 1]
```diff
@@ -43,3 +43,12 @@ export function async(fn: Function): (done: any) => any {
   // newest version of zone.js(0.8.25)
   return asyncFallback(fn);
 }
+
+/**
+ *
```
Comment (mhevery) [inline packages/core/testing/src/async.ts:48]: Needs docs of how to use. (Even if it says "see ...")

File: packages/core/testing/src/async.ts [hunk 2]
```diff
@@ -43,3 +43,12 @@ export function async(fn: Function): (done: any) => any {
...
+ * @publicApi
+ * @deprecated use asyncTest()
```
Comment (mhevery) [inline packages/core/testing/src/async.ts:50]:
```suggestion
 * @deprecated use `asyncTest()` (expected removal in v12)
```

File: packages/core/testing/src/async.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ import {asyncFallback} from './async_fallback';
  *
  * @publicApi
  */
-export function async(fn: Function): (done: any) => any {
+export function asyncTest(fn: Function): (done: any) => any {
```
Comment (mhevery) [inline packages/core/testing/src/async.ts:28]: `<bikeshedding>`
- `waitForAsync` Describes what the function does. 
- `waitForAsyncToComplete` Describes what the function does in a more complete sentence. 

`</bikeshedding>`

File: packages/core/testing/src/fake_async.ts [hunk 4]
```diff
@@ -43,14 +43,22 @@ export function resetFakeAsyncZone(): void {
...
     return fakeAsyncTestModule.fakeAsync(fn);
   } else {
     return fakeAsyncFallback(fn);
   }
 }
 
+/**
+ *
+ * @publicApi
+ * @deprecated use fakeAsyncTest()
```
Comment (mhevery) [inline packages/core/testing/src/fake_async.ts:57]:
```suggestion
 * @deprecated use `fakeAsyncTest()` (expected removal in v12)
```

File: packages/core/testing/src/fake_async.ts [hunk 5]
```diff
@@ -43,14 +43,22 @@ export function resetFakeAsyncZone(): void {
...
+export function fakeAsyncTest(fn: Function): (...args: any[]) => any {
   if (fakeAsyncTestModule) {
     return fakeAsyncTestModule.fakeAsync(fn);
   } else {
     return fakeAsyncFallback(fn);
   }
 }
 
+/**
+ *
```
Comment (mhevery) [inline packages/core/testing/src/fake_async.ts:55]: Needs JSDOC

File: packages/core/testing/src/fake_async.ts [hunk 6]
```diff
@@ -43,14 +43,22 @@ export function resetFakeAsyncZone(): void {
  *
  * @publicApi
  */
-export function fakeAsync(fn: Function): (...args: any[]) => any {
+export function fakeAsyncTest(fn: Function): (...args: any[]) => any {
```
Comment (mhevery) [inline packages/core/testing/src/fake_async.ts:46]: Not really convinced that renaming `fakeAsync` to `fakeAsyncTest` improves readability.

**Review by mhevery — CHANGES_REQUESTED** [review]

**Comment by mhevery** [comment]
- [ ] The backwards compatible rename can land in patch release. 
- [ ] The automatic migration tool should land before the major release. 
- [ ] Need to discuss with leads if this kind of rename is what we want.

File: packages/core/testing/src/async.ts (hunk 3 above)
Comment (JiaLiPassion) [reply packages/core/testing/src/async.ts:28]: Got it, I will rename it to `waitForAsync`.

**Comment by petebacondarwin** [comment]
I removed the review label and added cleanup because a) it has conflicts and needs a rebase; b) we have agreed (I believe) to use `waitForAsync ` and keep `fakeAsync`.

**Comment by JiaLiPassion** [comment]
@petebacondarwin , I have rebased and rename `async` to `waitForAsync`, also revert `fakeAsync`. Please review, thanks.

**Comment by petebacondarwin** [comment]
I'm afraid the AIO test code needs reformatting

File: packages/core/testing/src/async.ts (hunk 2 above)
Comment (JiaLiPassion) [reply packages/core/testing/src/async.ts]: Got it, updated, tahnks.

File: packages/core/testing/src/async.ts (hunk 1 above)
Comment (JiaLiPassion) [reply packages/core/testing/src/async.ts]: Added JSDoc `@see {@link waitForAsync}`

**Comment by JiaLiPassion** [comment]
@petebacondarwin , yeah, I updated the format and commit again, please review, thanks.

**Review by petebacondarwin — APPROVED** [review]
Reviewed-for: fw-upgrade

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:324084689:BASE:324084717:1596144868193:45b9734f)

**Review by mhevery — APPROVED** [review]
LGTM
- [ ] Sorry to be a pain, but could you break the PR into two. One for the change, and one to move `async` to `waitForAsync` We had it broken up but somehow it gote merged.

Reviewed-for: global-approvers
Reviewed-for: fw-core

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:324084689:BASE:324276416:1596228236860:99c16274)

**Comment by mhevery** [comment]
merge-assistance: global-approval
